File: src/lib/events/event-registry.ts

type RegisteredEventFn<Evt = void> = {
  id: string;
  fn: (evt: Evt) => void;
};

export class EventRegistry<Evt = void> {
  private eventFnMap: Map<string, RegisteredEventFn<Evt>>;
  private evtIdCounter: number;
  constructor() {
    this.eventFnMap = new Map();
    this.evtIdCounter = 0;
  }
  register(fn: (evt: Evt) => void): () => void {
    let evtId = this.getNextEvtId();
    this.eventFnMap.set(evtId, {
      id: evtId,
      fn,
    });
    return () => {
      this.eventFnMap.delete(evtId);
    };
  }
  fire(evt: Evt) {
    let evtIds = [ ...this.eventFnMap.keys() ];
    for(let i = 0; i < evtIds.length; ++i) {
      let currId = evtIds[i];
      let regEvt = this.eventFnMap.get(currId);
      regEvt?.fn(evt);
    }
  }
  eventFnCount(): number {
    return this.eventFnMap.size;
  }
  private getNextEvtId(): string {
    return `${this.evtIdCounter++}`;
  }
}

File: src/lib/events/eventregistry.go
package events

type registeredEventFn[T any] struct {
	Id int
	Fn func(evt T)
}

type EventRegistry[T any] struct {
	eventFnMap   map[int]registeredEventFn[T]
	evtIdCounter int
}

func NewEventRegistry[T any]() EventRegistry[T] {
	return EventRegistry[T]{
		eventFnMap: make(map[int]registeredEventFn[T]),
	}
}

func (er *EventRegistry[T]) Register(fn func(evt T)) func() {
	evtId := er.getNextEvtId()
	regEvt := registeredEventFn[T]{
		Id: evtId,
		Fn: fn,
	}
	er.eventFnMap[evtId] = regEvt
	offCb := func() {
		delete(er.eventFnMap, evtId)
	}
	return offCb
}

func (er *EventRegistry[T]) Fire(evt T) {
	for _, regEvt := range er.eventFnMap {
		regEvt.Fn(evt)
	}
}

func (er *EventRegistry[T]) getNextEvtId() int {
	nextId := er.evtIdCounter
	er.evtIdCounter = er.evtIdCounter + 1
	return nextId
}
